src/embeddings/tools/sgns.cpp

```diff
@@ -0,0 +1,607 @@
+                        int64_t& word_counter,
+                        std::default_random_engine& engine,
+                        std::uniform_int_distribution<std::size_t>& next_random)
+    {
+        while (*stream)
+        {
+            const std::string word = stream->next();
+            auto i = vocab_.table.find(word);
+
+            // Ignore out-of-vocabulary words.
+            if (i != vocab_.table.end())
+            {
+                ++word_counter;
+
+                if (subsample_threshold_ > 0)
+                {
+                    const auto count = vocab_.vector[i->value()].count;
+                    const float ran
+                        = ((float)sqrt(count / (float)(subsample_threshold_
+                                                       * vocab_.total_count))
+                           + 1)
+                          * (float)(subsample_threshold_ * vocab_.total_count)
+                          / count;
+                    if ((ran < (next_random(engine) & 0xFFFF) / 65536.0f))
+                    {
+                        continue;
+                    }
+                }
+
+                window.push_back(i->value());
+
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    // Load the vocabulary from a file.
+    sgns_vocab load_vocab(const std::string& filename) const
+    {
+        if (!filesystem::file_exists(filename))
+        {
+            LOG(fatal) << "Vocabulary has not yet been generated, please "
+                          "do this before learning word embeddings"
+                       << ENDLG;
+            throw sgns_exception{"no vocabulary file found in " + prefix_};
+        }
+
+        std::ifstream input{filename, std::ios::binary};
+        const auto size = io::packed::read<uint64_t>(input);
+
+        sgns_vocab vocab{};
+
+        printing::progress progress{" > Loading vocab: ", size};
+
+        for (uint64_t tid{0}; tid < size; ++tid)
+        {
+            progress(tid);
+            const auto word = io::packed::read<std::string>(input);
+            const auto count = io::packed::read<uint64_t>(input);
+            vocab.table[word] = tid;
+            vocab.vector.push_back(sgns_word{word, count});
+            vocab.total_count += count;
+        }
+
+        progress.end();
+
+        const uint64_t bytes_used
+            = (vocab.vector.size() * sizeof(sgns_vocab_vector::value_type))
+              + vocab.table.bytes_used();
+
+        LOG(info) << "Loaded vocabulary of size " << vocab.table.size()
+                  << " occupying " << printing::bytes_to_units(bytes_used)
+                  << ENDLG;
+
+        if ((max_ram_ <= bytes_used)
+            || ((max_ram_ - bytes_used) < (1024 * 1024)))
+        {
+            throw sgns_exception{"RAM limit too restrictive"};
+        }
+
+        return vocab;
+    }
+
+    // Create the noise distribution from which negative word samples are
+    // randomly drawn.
```

Comment:
Can you increase the verbosity of this comment just a little bit? This doesn't return a distribution; it returns a vector of length `size` whose elements are present in proportion to their noise sampling probability. Samples can then be drawn by sampling an index into the vector uniformly at random.

(I'm also curious if this even matters, performance wise. But that's how they do it, so presumably there's an improvement...)